README.md
# Ansible Webhost

This package contains an Ansible playbook for configuring a Debian host with a combination of Nginx, PHP, and MySQL. The host will have UFW/IPTables running, and websites can be deployed via Git automatically.

## Requirements

* Ubuntu 14.04 LTS server
* Root (or user with sudo) permissions

## To Do

* Set up a root MySQL password for each server, instead of reusing a single password
* [Potentially Solved] - Deploy SSL certificate for Nginx/Haproxy when deploying a site with SSL
* Hostname is not set correctly on Ubuntu servers - Can't resolve itself
* Enable HSTS in Nginx and HAProxy for SSL enabled websites

## Instructions

* Install [geerlingguy/ansible-role-mailhog](https://github.com/geerlingguy/ansible-role-mailhog) via Ansible Galaxy:
  * `ansible-galaxy install geerlingguy.mailhog`
* Copy `inventory/example.config.php` to `inventory/config.php`, and configure it with your servers
* Add a public key for every administrative user into roles/ssh/files/keys/[username].pub
  * Root access and Password Authentication will be disabled, so be sure this is a working public key or you may lock yourself out.
* The first time you deploy the webserver with this script, use the command:
  * `ansible-playbook main.yml -e "ansible_ssh_user=root default_password=ThisIsTheDefaultPassword" --ask-pass`
  * This will create the user accounts, and then fail to process the remaining roles. This is intentional, as we disable the root user.
* After the initial user configuration, you can use:
  * `ansible-playbook main.yml --ask-sudo-pass`


release_notes.md
# Release 1.3

## New features

* Memcached
* Load balancing
* MySQL Master/Slave configuration
* MySQL Database and User creation
* PHP Pool configuration
* Customizable Nginx/PHP configurations for each website

## Breaking Changes

None expected


inventory/defaults.php
<?php

function getHostDefaults($type, $host) {
  global $generic_host, $default_host_types, $groups;
  $default_group = empty($default_host_types[$type]) ? array() : $default_host_types[$type];
  $custom_group = empty($groups[$type]) ? array() : $groups[$type];
  return array_replace_recursive($default_group, $custom_group, $host);
}

$config = array(
  '_meta' => array(
  ),
);

$generic_host = array(
  'administrators' => array(),
  'ssh_allow_nonadmins' => array(),
  'management_ips' => array(),
);

$default_host_types = array(
  'development' => array(
    'websites' => array(
      '__hostname__' => array(
        'git' => array(
          'repo' => 'git@gitlab.com:simple-simple-advertising/devbox-landing-page.git',
        ),
      ),
    ),
    'nginx' => true,
    'haproxy' => false,
    'varnish' => false,
    'memcached' => FALSE,
  ),

  'production' => array(
    'nginx' => true,
    'nginx_port' => 8081,
    'nginx_spdy_port' => 8082,

    'haproxy' => true,
    'haproxy_port' => 80,
    'haproxy_ssl_port' => 443,

    'varnish' => true,
    'varnish_port' => 8080,

    'memcached' => FALSE,
  ),

  'webserver' => array(
    'nginx_host' => '127.0.0.1',
    'websites' => array(
      'phpmyadmin.__hostname__' => array(
        'git' => array(
          'repo' => 'https://github.com/phpmyadmin/phpmyadmin.git',
          'branch' => 'MAINT_4_0_10',
        ),
        'type' => 'php',
        'nginx_tweak' => 'client_max_body_size 1024m;',
        'loadbalanced' => FALSE,
      ),
    ),
    'loadbalancer' => array('127.0.0.1'),
  ),

  'database' => array(
    'database' => array(
      'exposed' => FALSE,
    ),
  ),
);


inventory/inventory.php
#!/usr/bin/env php
<?php
require 'config.php';

if (isset($config_raw) === false || is_array($config_raw) === false) {
  $config_raw = array();
}

$config_raw = $config;

require 'defaults.php';

$options = getopt('v', array('host:', 'list'));

$pretty = (isset($options['v']) === true) ? JSON_PRETTY_PRINT : NULL;

if (empty($options['host']) === false) {
  $host = $options['host'];

  if (empty($config_raw[$host]) === true) {
    echo json_encode(array());
    return;
  }
  echo json_encode(processHostVars($host, $config_raw[$host]), $pretty);
  return;
}

function processHostVars($host, $settings) {
  global $config;

  // Is this server a webserver?
  if (isset($settings['webserver']) === true && $settings['webserver'] === true) {
    $config['webserver']['hosts'][] = $host;
    $settings = getHostDefaults('webserver', $settings);
  }

  // Is this server a database?
  if (isset($settings['database']) === true && $settings['database'] !== FALSE) {
    $config['database']['hosts'][] = $host;
    $settings = getHostDefaults('database', $settings);
  }

  // Set this into the correct environment hostgroup
  if (empty($settings['environment']) === false) {
    $config[$settings['environment']]['hosts'][] = $host;
    $settings = getHostDefaults($settings['environment'], $settings);
  }

  // Ensure this host's hostname is set
  if (empty($settings['hostname']) === true) {
    $settings['hostname'] = $host;
  }

  foreach ($settings['websites'] as $domain => &$website) {
    // If the host config has disabled this site, remove it
    if ($website === false) {
      unset($settings['websites'][$domain]);
    }

    // Handle any hostname replacement on the domain name
    $website['domain'] = str_replace('__hostname__', $settings['hostname'], $domain);

    if (empty($website['ssl']) === false) {
      $settings['host_ssl'] = true;
    }

    if (empty($website['use_www']) === TRUE) {
      $website['use_www'] = (strtolower(substr($website['domain'], 0, 4)) === 'www.');
    }

    // Strip www. off the domain before we set everything up.
    if (strtolower(substr($website['domain'], 0, 4)) === 'www.') {
      $website['domain'] = substr($website['domain'], 4);
    }

    // Set a user for this site to run as
    $website['user'] = substr($website['domain'], 0, 32);

    if ($website['use_www'] === FALSE) {
      $website['domain_redirected'] = 'www.' . $website['domain'];
    } else {
      $website['domain_redirected'] = $website['domain'];
      $website['domain'] = 'www.' . $website['domain'];
    }
  }
  $settings['websites'] = array_values($settings['websites']);

  return $settings;
}

foreach ($config_raw as $host => $value) {
  $config['_meta']['hostvars'][$host] = processHostVars($host, $value);
}

if (isset($options['list']) === true) {
  unset($config['_meta']);
}

echo json_encode($config, $pretty);


roles/ansible-master/files/bootstrap.js
#!/usr/bin/env node
var prompt = require('prompt')
  , sys = require('sys')
  , exec = require('child_process').exec
  , spawn = require('child_process').spawn
  , fork = require('child_process').fork
  , temp = require('temp')
  , fs = require('fs');

//temp.track();
prompt.start();

var schema = {
  properties: {
    email: {
      pattern: /^.*@simplesimple\.ca$/,
      message: 'Email must be @simplesimple.ca',
      required: true,
    },
    password: {
      hidden: true,
      required: true,
    },
  },
};

prompt.get(schema, function (err, result) {
  if (err) { console.log(err); return; }
  result.username = result.email.match(/^(.*)@simplesimple\.ca$/)[1];
  result.hostname = 'dev-' + result.username;
  result.client_ip = process.env.SSH_CLIENT.split(' ')[0];
  setup_playbook(result);
});

var setup_playbook = function setup_playbook(settings) {
  temp.open('ansible-bootstrap', function (err, playbook) {
    fs.write(playbook.fd, settings.client_ip +
      ' hostname=' + settings.hostname +
      ' email=' + settings.email +
      ' username=' + settings.username);
    fs.close(playbook.fd, function (err) {
      run_playbook(playbook.path, settings);
    });
  });
};

var run_playbook = function run_playbook(inventory_file, settings) {
  var known_hosts = fs.openSync('/home/ansible-bootstrap/.ssh/known_hosts', 'a');
  spawn('ssh-keyscan', ['-H', settings.client_ip], { stdio: [ 'ignore', known_hosts, 'ignore' ]})
    .on('close', function (code) {
      spawn('ansible-playbook', ['-i', inventory_file, '--extra-vars', '"password=$(openssl\\ passwd\\ -1\\ -salt\\ $(openssl\\ rand\\ -base64\\ 6)\\ ' + settings.password + ')"', '-u', 'vagrant', '/etc/ansible/bootstrap.yml'], { stdio: 'inherit' })
        .on('close', function (code) {
          var hosts = fs.readFileSync('/etc/ansible/hosts', { encoding: 'utf-8' });
          if ((new RegExp('\\[devs\\](?:\\n.+)*?\\n' + settings.hostname)).test(hosts) === false) {
            exec('sudo sed -i "s/\\(\\[devs\\]\\)/\\1\\n' + settings.hostname + '/" /etc/ansible/hosts', function (err, stdout, stderr) {
              console.log('\nTo SSH to the server, please use `ssh ' + settings.username + '@' + settings.hostname + '`\n\n');
            });
          } else {
            console.log('\nTo SSH to the server, please use `ssh ' + settings.username + '@' + settings.hostname + '`\n\n');
          }
      //exec('./cleanup.sh ' + settings.hostname);
        });
    });
};


roles/apache2/files/sites/landingpage/index.php
<?php
if (strpos($_SERVER['REMOTE_ADDR'], '10.0.0') !== 0) {
  header('HTTP/1.0 404 Not Found');
  echo file_get_contents('./404.html');
  return;
}
?>
<!DOCTYPE html>
<html class="no-js">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
    <title>Simple Simple Staging</title>
    <meta name="description" content="">
    <meta name="viewport" content="width=device-width">

    <link rel="stylesheet" href="http://landingpage.<?php echo $_SERVER['HTTP_HOST']; ?>/css/normalize.min.css">
    <link rel="stylesheet" href="http://landingpage.<?php echo $_SERVER['HTTP_HOST']; ?>/css/main.css">

    <script src="http://landingpage.<?php echo $_SERVER['HTTP_HOST']; ?>/js/vendor/modernizr-2.6.2.min.js"></script>
  </head>
  <body>
    <h1>Hello Simple Simple!</h1>

    <ul>
<?php
$dir = '/var/www/shared/';
$sites = array();
if (is_dir($dir)) {
  if ($dh = opendir($dir)) {
    while (($file = readdir($dh)) !== false) {
      if (filetype($dir . $file) === 'dir' && $file !== '.' && $file !== '..') {
        $sites[$file] = "      <li><a href='http://$file.{$_SERVER['HTTP_HOST']}/'>$file</a></li>";
      }
    }
    closedir($dh);
  }
}
ksort($sites);
echo implode("\n", array_values($sites));
?>
    </ul>

    <script src="//ajax.googleapis.com/ajax/libs/jquery/1.10.1/jquery.min.js"></script>
    <script>window.jQuery || document.write('<script src="http://landingpage.<?php echo $_SERVER['HTTP_HOST']; ?>/js/vendor/jquery-1.10.1.min.js"><\/script>')</script>
    <script src="http://landingpage.<?php echo $_SERVER['HTTP_HOST']; ?>/js/main.js"></script>
  </body>
</html>


inventory/example.config.php
<?php
/**
 * Configure settings that will be applied to any servers in this a group.
 */
$groups = array(
  'development' => array(
    // Users that are allowed to manage these servers.
    // These users must have a matching public key in roles/ssh/files/keys.
    // eg: roles/ssh/files/keys/spenser.pub for spenser.
    'administrators' => array('spenser'),
    // Any IPs that are allowed to manage these servers.
    'management_ips' => array('10.0.0.0/24'),
  ),

  'production' => array(
    // Users that are allowed to manage these servers.
    // These users must have a matching public key in roles/ssh/files/keys.
    // eg: roles/ssh/files/keys/spenser.pub for spenser.
    'administrators' => array('spenser'),
    // Other users that are allowed to SSH in, and are not administrators.
    'ssh_allow_nonadmins' => array('automatedbackups'),
    // Any IPs that are allowed to manage these servers.
    'management_ips' => array('184.71.62.238'),
  ),
);

/**
 * Configure individual servers, keyed by the name you SSH to the server with.
 *
 * No options are required, and will default to deploying a barebones server.
 */
$config = array(
  'hosting.example.com' => array(
    // Does this server have a different hostname than you connect to it with?
    'hostname' => 'example.com',

    // Is this server a webserver?
    'webserver' => TRUE,
    // Does this server have a database server on it?
    'database' => FALSE,
    // Does this server have memcached on it?
    'memcached' => TRUE,
    // Does this server have haproxy on it?
    'haproxy' => TRUE,
    // Does this server have Varnish on it?
    'varnish' => TRUE,
    // Is this a Production or Development server?
    'environment' => 'production',

    // Configure ports for Haproxy, Varnish, and Nginx.
    'nginx' => true,
    'nginx_port' => 8081,
    'nginx_spdy_port' => 8082,
    'haproxy' => true,
    'haproxy_port' => 80,
    'haproxy_ssl_port' => 443,
    'varnish' => true,
    'varnish_port' => 8080,

    // Where can this haproxy server find its Nginx host?
    'nginx_host' => '127.0.0.1',

    // Override the users that are allowed to manage this server.
    // This has the same setup as the defaults in $groups.
    'administrators' => array('spenser', 'john'),
    // Override the IPs that are allowed to SSH to this server.
    'management_ips' => array('192.168.0.0/24'),
    // Override users that are allowed to SSH in, and are not administrators.
    'ssh_allow_nonadmins' => array('guest'),

    // Options for each website that is deployed to this server.
    'websites' => array(
      // __hostname__ will be replaced with the hostname of this server.
      // This allows for easily configuring subdomains.
      '__hostname__' => array(
        // Should this website be automatically cloned from a git repository?
        'git' => array(
          // The repository to clone. The public key in roles/ssh/files/keys
          // must have access to this repository.
          'repo' => 'git@gitlab.com:simple-simple-advertising/bowvalleycollege.git',
          // Which branch should be automatically checked out?
          'branch' => 'master',
        ),

        // What type of site is this? (drupal6, drupal7, wordpress, static, php)
        'type' => 'drupal7',
        // Which SSL certificate should be installed for this website?
        // This will automatically force all traffic through SSL.
        'ssl' => 'ssl_certs/example.com/example.com.crt',

        // Customize the Nginx Site configuration.
        'nginx_tweak' => 'client_max_body_size 64m;',

        // Customize the PHP configuration for this website.
        'php_tweak' => array(
          'php_admin_value[post_max_size] = 64M',
          'php_admin_value[upload_max_filesize] = 64M',
        ),
        // Customize the PHP Pool processes.
        'php_process_max_total' => 15,
        'php_process_initial' => 5,
        'php_process_min_spare' => 5,
        'php_process_max_spare' => 10,
        'php_process_max_requests' => 500,

        // Which path should HAProxy check to see if the server is online?
        'httpcheck' => '/install.php',
        // Should this site allow robots to crawl it?
        'allow_robots' => FALSE,

        // Should loadbalancing be enabled for this website?
        'loadbalanced' => TRUE,
      ),
    ),

    // Is this webserver loadbalanced with another server?
    // If this is enabled, all websites will need to exist on both servers.
    'loadbalancer' => array(
      '192.168.0.5',
      '192.168.0.6',
    ),

    // If memcached is installed, what servers should it be exposed to?
    // This will typically match the loadbalancer list.
    'memcached' => array(
      'exposed' => array(
        '192.168.0.5',
        '192.168.0.6',
      ),
    ),
  ),

  'db1.example.com' => array(
    'environment' => 'production',

    // database can be a boolean for basic DB support, or an array of
    // configuration options, which allows provisioning dbs, users, and
    // master/slave relationships.
    'database' => array(
      // What servers is this database exposed to? (array|bool)
      'exposed' => array(
        '192.168.0.5',
        '192.168.0.6',
      ),
      // Which users should be provisioned on this database server?
      'users' => array(
        array(
          'user' => 'example.com',
          'host' => '192.168.0.5',
          'password' => 'password1234forWeb1',
          'privileges' => 'example.com.*:ALL',
        ),
        array(
          'user' => 'example.com',
          'host' => '192.168.0.6',
          'password' => 'password1234forWeb2',
          'privileges' => 'example.com.*:ALL',
        ),
        array(
          'user' => 'read-example.com',
          'host' => '192.168.0.5',
          'password' => 'readonlyForWeb1',
          'privileges' => 'example.com.*:SELECT',
        ),
      ),
      // Which databases should be provisioned on this server?
      'databases' => array(
        'example.com',
      ),
    ),
    // If this is a MySQL master, configure the slaves that can access it.
    'mysql_master' => array(
      'slaves' => array(
        '192.168.0.11' => 'passwordForSlaveUser',
      ),
    ),
  ),

  'db2.example.com' => array(
    'webserver' => FALSE,
    'database' => array(
      // What servers is this database exposed to? (array|bool)
      'exposed' => array(
        '192.168.0.5',
        '192.168.0.6',
      ),
    ),
    // Configure this server as a MySQL slave.
    'mysql_slave' => array(
      'master' => '192.168.0.10',
      'password' => 'passwordForSlaveUser',
      // This ID should be >=2, and should increment for every additional slave.
      'id' => 2,
    ),
    'environment' => 'production',
  ),
);
